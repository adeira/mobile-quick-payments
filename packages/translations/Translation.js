// @flow

import * as React from 'react';
import { Text } from 'react-native'; // eslint-disable-line no-restricted-imports
import { FormattedMessage } from 'react-intl';
import { type TranslationKeys } from 'mobile-quick-payments-translations';

type CommonProps = {|
  testID?: string,
|};

type Props =
  | {|
      id: TranslationKeys,
      ...CommonProps,
    |}
  | {|
      passThrough: string,
      ...CommonProps,
    |};

export default class Translation extends React.PureComponent<Props> {
  render = () => {
    const p = this.props;

    if (p.id !== undefined && p.passThrough !== undefined) {
      throw new Error(
        "You can use only 'id' or 'passThrough' property in translations but not both.",
      );
    }

    if (p.passThrough !== undefined) {
      return <Text>{p.passThrough}</Text>;
    }

    return (
      <FormattedMessage id={p.id}>
        {translatedText => <Text>{translatedText}</Text>}
      </FormattedMessage>
    );
  };
}
