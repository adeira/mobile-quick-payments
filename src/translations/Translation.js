// @flow

import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import { type TranslationKeys } from 'mobile-quick-payments-translations';

type Props = {|
  id: TranslationKeys,
|};

export default ({ id }: Props) => <FormattedMessage id={id} />;
