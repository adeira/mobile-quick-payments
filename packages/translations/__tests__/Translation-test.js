// @flow

import Translation from '../Translation';

it('works with translation ID property', () => {
  // this will call the underlying translation engine
  expect(
    // $FlowExpectedError: string "ID test" is not a valid translation key
    new Translation({
      id: 'ID test',
    }).render(),
  ).toMatchSnapshot();
});

it('works with `passThrough` property', () => {
  // this will just return Text tag directly
  expect(
    new Translation({
      passThrough: 'passThrough test',
    }).render(),
  ).toMatchSnapshot();
});

it('throws exception for unclear properties', () => {
  expect(() =>
    // $FlowExpectedError: ID and passThrough at the same time is not a valid type
    new Translation({
      id: 'ID test',
      passThrough: 'passThrough test',
    }).render(),
  ).toThrowErrorMatchingSnapshot();
});

test.each([['string'], [123], [null]])(
  'works with almost any type of `passThrough` property - testing: %p',
  value => {
    expect(
      new Translation({
        passThrough: value,
      }).render(),
    ).toMatchSnapshot();
  },
);
