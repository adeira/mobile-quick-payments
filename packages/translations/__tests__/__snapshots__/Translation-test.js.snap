// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws exception for unclear properties 1`] = `"You can use only 'id' or 'passThrough' property in translations but not both."`;

exports[`works with \`passThrough\` property 1`] = `
<Text>
  passThrough test
</Text>
`;

exports[`works with almost any type of \`passThrough\` property 1`] = `
<Text>
  string
</Text>
`;

exports[`works with almost any type of \`passThrough\` property 2`] = `
<Text>
  123
</Text>
`;

exports[`works with almost any type of \`passThrough\` property 3`] = `<Text />`;

exports[`works with almost any type of \`passThrough\` property 4`] = `<Text />`;

exports[`works with translation ID property 1`] = `
<FormattedMessage
  id="ID test"
  values={Object {}}
>
  [Function]
</FormattedMessage>
`;
